### PR TITLE
feat: add rootSpan.createChildSpan and change none CLS semantics

### LINKED ISSUE #719: Option to disable automatic tracing
Hello,

Is there any way to disable automatic tracing altogether? It would be nice if this module exposed a simple span tracing API. 

Thanks

### PR DESCRIPTION
**Pull request by kjin**
Fixes #719

This PR does two things:

* In #728, a new `config.clsMechanism` option was added where, when set to `'none'`, a barebones CLS mechanism would be used where when a root span was started, it became the only possible active root span context until it was ended. This PR changes this so that root span context will _never_ be written when `clsMechanism` is set to `'none'`. Instead, the root span context always has a default value of `UNTRACED`. This affects the behavior of the custom tracing API in the following ways:
  * `traceApi.runInRootSpan` will _always_ return a root span (unless disallowed by the trace policy), even if run within a function passed to another call to `runInRootSpan`.
  * `traceApi.createChildSpan` will _always_ return an `UNTRACED` child span. This is because root spans can no longer be automatically determined from the root context.
* To address the second difference above, there is now a `rootSpan.createChildSpan` which takes the same options as `traceApi.createChildSpan`. When `clsMechanism` is set to `'none'`, the latter will not work, but the former will; when `clsMechanism` is set to/defaults to `'auto'`, they should behave the same way.

### CODE REVIEW

**Comment by tjmehta** [comment]
Wow you are fast @kjin! I am out of town and will check this out tomorrow

File: scripts/check-install.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export async function checkInstall() {
   await spawnP('npm', ['init', '-y'], {
     cwd: installDir
   });
-  await spawnP('npm', ['install', 'typescript', '@types/node', tgz[0]], {
+  await spawnP('npm', ['install', 'typescript', '@types/node@9', tgz[0]], {
```
Comment (jinwoo) [inline scripts/check-install.ts]: Why is this needed?

File: src/cls.ts [hunk 2]
```diff
@@ -137,23 +139,25 @@ export class TraceCLS implements CLS<RootContext> {
   }
 
   enable(): void {
-    if (!this.enabled) {
+    // if this.CLSClass = DefaultCLS, the user specifically asked not to use
```
Comment (jinwoo) [inline src/cls.ts]: So `DefaultCLS` is a CLS that does nothing, am I correct? The name seems confusing then. What about `NullCLS`?

File: src/span-data.ts [hunk 3]
```diff
@@ -18,7 +18,8 @@ import * as crypto from 'crypto';
 import * as util from 'util';
 
 import {Constants, SpanDataType} from './constants';
-import {SpanData as SpanData} from './plugin-types';
+import * as types from './plugin-types';
+import {SpanData, SpanOptions} from './plugin-types';
```
Comment (jinwoo) [inline src/span-data.ts]: You seem to use only `types.SpanData` in this file. Why separately import these BTW?

File: src/span-data.ts [hunk 4]
```diff
@@ -120,6 +121,16 @@ export class RootSpanData extends BaseSpanData {
     this.span.kind = SpanKind.RPC_SERVER;
   }
 
+  createChildSpan(options: SpanOptions): SpanData {
+    options = options || {name: ''};
```
Comment (jinwoo) [inline src/span-data.ts]: Is `options` optional? Why do you need this line?

Comment (kjin) [reply src/span-data.ts]: They're optional, but in the absence of type checking it prevents an error from being thrown when options aren't passed in. It's sort of a relic of older versions of this module -- I added it to keep it consistent with the same precondition check in `traceApi.createChildSpan`.

Comment (jinwoo) [reply src/span-data.ts]: Erm... I'm not familiar with the history of this module, but why not mark it as optional then?

Comment (kjin) [reply src/span-data.ts]: The "history of this module" is just that it was once written in JavaScript :)

I've marked `options` as optional.

File: src/cls.ts [hunk 5]
```diff
@@ -137,23 +139,25 @@ export class TraceCLS implements CLS<RootContext> {
   }
 
   enable(): void {
-    if (!this.enabled) {
+    // if this.CLSClass = NoneCLS, the user specifically asked not to use
```
Comment (jinwoo) [inline src/cls.ts]: s/NoneCLS/NullCLS

File: src/span-data.ts [hunk 6]
```diff
@@ -18,7 +18,8 @@ import * as crypto from 'crypto';
 import * as util from 'util';
 
 import {Constants, SpanDataType} from './constants';
-import {SpanData as SpanData} from './plugin-types';
+import * as types from './plugin-types';
```
Comment (jinwoo) [inline src/span-data.ts]: Why not just add `RootSpanData` to the import statement below? You don't seem to use anything other than those three (SpanData, SpanOptions, and RootSpanData), do you?

Comment (kjin) [reply src/span-data.ts]: It conflicts with `RootSpanData` defined in this file. I couldn't think of any way to avoid this at the time.

Come to think of it, OpenCensus calls the write-only object "Span" (and SpanData refers to the read-only object given to exporters). I'll add a TODO to change this.

**Comment by kjin** [comment]
@tjmehta Were you able to get a chance to look at the new API? (I understand that it's a sizeable PR -- I think the user-visible changes should most be encapsulated in [this file](https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/pull/731/files#diff-0083d751b2c247998a6ebdda46fffa61))
